text/0140-record-feature-flags-on-error.md

```diff
@@ -0,0 +1,45 @@
+- Start Date: 2024-09-04
+- RFC Type: feature
+- RFC PR: https://github.com/getsentry/rfcs/pull/140
+- RFC Status: draft
+
+# Summary
+
+As feature flags are evaluated in a customer's application we will collect those flags and their results, store them in-memory, and push the values to Sentry when an error occurs. The flags can be used to further understand what went wrong during the session.
+
+# Motivation
+
+To enable users to debug application errors with a complete picture of their application's state.
+
+# Options Considered
+
+We will collect flag evaluations and hold them in-memory.  On error event the flags will be placed into the `contexts` object on the `event` and sent to Sentry.
+
+## Transport
+
+The flags will be represented by this data structure during transport. The `flag` key is the name of the flag and the `result` key is the evaluation result returned by the customer's application.
+
+```json
+{
+    "contexts": {
+        "flags": [
+            {"flag": "abc", "result": true},
+            {"flag": "def", "result": false}
+        ]
+    }
+}
+```
+
+## Public Interface
+
+The SDK will expose one new public method `set_flag`.  Similar to `set_tag` or `set_user`, the `set_flag` method sets a key, value pair (representing the flag's name and its evaluation result) into an internal SDK structure. On error, that structure is serialized and appended to the event body as described in the previous section.
```

Comment:
Not a big fan of another method. With spans, we will unify to one single `setAttributes` method. For flags, using `setContext('flags', ...)` should suffice.

Reply:
How would someone incrementally add flags as they're evaluated?  How would we manage the capacity on the flags key?

Reply:
I'm not a fan of magic strings tbqh. If we're making flags a first class citizen, it's beneficial to have a direct API for it. Also for discoverability.

Is the concern here bundle size on JS? Or just "too many functions/complexity"?

Reply:
My main concern here is to add new methods before the new feature was fully evaluated.
If we scope the initial implementation to one or two SDKs and trail it for a while behind some feature experimental feature flag, we can proceed. I'm not too worried about the capacity at first, though merging values is a good argument against using `setContext()`.

Reply:
> My main concern here is to add new methods before the new feature was fully evaluated.

That's fair. No need to push for publish APIs while we're in alpha/beta phase. 

The RFC is proposing a solution long term though, in a world where we're not canceling this project, the final API proposal is what we agree on in this document